### PR TITLE
Remote data improvements

### PR DESCRIPTION
**Pull request by blazyliszek**
### Description

We used this RemoteData pattern across different projects, so we learned whats good and bad about it. This is just slight refactor / fixes for some of its problems.

### Changes

- failure type might contain data now (in case our data flow is Success -> Refreshing -> Failure we don't want to erase it)
- refactor (var names / var name cases /  file placement) 
- added Pending type

### Considerations

Our RemoteData still has some flaws which I listed [here](https://nozbe.app/teams/Xi3SI50NCpPkVxNz/task_id/3sAjCbeo8gJzHjcA) if you're interested.

### CODE REVIEW

File: template/src/utils/api.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+export const isRefreshing = <T, E>(remoteData: RemoteData<T, E>): remoteData is RefreshingType<T> =>
+  remoteData.state === RemoteDataStates.REFRESHING
+
+export const isFailure = <T, E>(remoteData: RemoteData<T, E>): remoteData is FailureType<E, T> =>
+  remoteData.state === RemoteDataStates.FAILURE
+
+export const isSuccess = <T, E>(remoteData: RemoteData<T, E>): remoteData is SuccessType<T> =>
+  remoteData.state === RemoteDataStates.SUCCESS
+
+export const hasData = <T, E>(
```
Comment (lukiwolski) [inline template/src/utils/api.ts:84]: what would be the usecase of this check?

Comment (blazyliszek) [reply template/src/utils/api.ts:84]: 1. lets say that we have object of type `RemoteData`. Until you do this check on it, typescript won't be happy if you try to access `data` property on it, saying that there is no `data` field on `NotRequested` / `LoadingType`, cause these types are part of the `RemoteData` union too.
2. Lets say you have `RemoteData` object and you want to display whatever it is and you don't care whether its state is a `success` / `refreshing` or `failure`.

Keep in mind its not a new code, but also happy to know if there is something in it that bothers you. It's not the first time when we want to fix it for good, but it still happen to come with some unresolved drawbacks.

Comment (lukiwolski) [reply template/src/utils/api.ts:84]: looks good, it wasnt bothering me, rather wanted to refamiliarize myself with the template :)

**Review by szymonkoper — APPROVED** [review]
👍

File: template/src/utils/api.ts (hunk 1 above)
Comment (lukiwolski) [reply template/src/utils/api.ts:84]: thinking about it now makes me wonder whether I shouldnt *blazenly* use this stuff in our project ;D

Comment (blazyliszek) [reply template/src/utils/api.ts:84]: I really like this concept, but on the other hand there are still some problems with it, that I didn't find solution for.
especially second one here -> https://nozbe.app/teams/Xi3SI50NCpPkVxNz/task_id/3sAjCbeo8gJzHjcA

**Pull request merged**